migrations/3_deploy_contracts.js

```diff
@@ -9,31 +11,104 @@ const IdentityVerifier = artifacts.require("IdentityVerifier")
-            return deployer.deploy(IdentityVerifier, Controller.address)
+    deployer.then(async () => {
+        const controller = await deploy(deployer, Controller)
+
+        const token = await deployAndRegister(deployer, controller, LivepeerToken, "LivepeerToken")
+        const minter = await deployAndRegister(deployer, controller, Minter, "Minter", config.minter.initialTokenSupply, config.minter.yearlyInflation)
+
+        let verifier
+
+        if (network === "development" || network === "testrpc" || network == "parityDev" || network === "gethDev") {
+            verifier = await deployAndRegister(deployer, controller, IdentityVerifier, "Verifier", controller.address)
         } else if (network === "lpTestNet") {
-            return deployer.deploy(LivepeerVerifier, Controller.address, config.verifier.solvers, config.verifier.verificationCodeHash)
+            verifier = await deployAndRegister(deployer, controller, LivepeerVerifier, "Verifier", controller.address, config.verifier.verificationCodeHash)
         } else {
-            return deployer.deploy(OraclizeVerifier, Controller.address, config.verifier.verificationCodeHash, config.verifier.gasPrice, config.verifier.gasLimit)
+            verifier = await deployAndRegister(deployer, controller, OraclizeVerifier, "Verifier", controller.address, config.verifier.verificationCodeHash, config.verifier.gasPrice, config.verifier.gasLimit)
         }
-    }).then(() => {
-        // Deploy upgradeable proxy target contracts
-        return deployer.deploy([
-            [BondingManager, Controller.address],
-            [JobsManager, Controller.address],
-            [RoundsManager, Controller.address]
-        ])
+
+        if (network === "development" || network === "testrpc" || network === "parityDev" || network == "gethDev" || network === "lpTestNet") {
+            const faucet = await deployAndRegister(deployer, controller, LivepeerTokenFaucet, "LivepeerTokenFaucet", token.address, config.faucet.requestAmount, config.faucet.requestWait)
+
+            await token.mint(faucet.address, new BigNumber(config.faucet.faucetAmount))
+
+            await Promise.all(config.faucet.whitelist.map(addr => {
+                return faucet.addToWhitelist(addr)
+            }))
+        }
+
+        const bondingManager = await deployProxyAndRegister(deployer, controller, BondingManager, "BondingManager", controller.address)
+        const jobsManager = await deployProxyAndRegister(deployer, controller, JobsManager, "JobsManager", controller.address)
+        const roundsManager = await deployProxyAndRegister(deployer, controller, RoundsManager, "RoundsManager", controller.address)
+
+        const owner = await controller.owner()
+
+        deployer.logger.log("Adding permissions...")
+
+        await controller.addPermission(owner, bondingManager.address, functionSig("setParameters(uint64,uint256,uint256)"))
```

Comment:
I could see a need for a future test that checks all of these hardcoded signatures against actual signatures in our compiled contract ABIs to report if any are not present. This could be an area of pain if we don't have checks, as developers would have to know to come in here to modify this if we ever make a signature change, and that wouldn't be obvious from a failing test.